pkg/middleware/allowed_hosts.go

```diff
@@ -0,0 +1,132 @@
+/****************************************************************************
+ * Copyright 2020, Optimizely, Inc. and contributors                        *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package middleware //
+package middleware
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+var errInvalidRequestHost = errors.New("invalid request host")
+
+// AllowedHosts returns a middleware function that rejects requests whose host value does not match any host in allowedHosts.
+// Request host is determined in the following priority order:
+// 1. X-Forwarded-Host header value
+// 2. Forwarded header host= directive value
+// 3. Host property of request (see Host under https://golang.org/pkg/net/http/#Request)
+func AllowedHosts(allowedHosts []string, allowedPort string, usingTLS bool) func(next http.Handler) http.Handler {
+	allowedMap := make(map[string]bool)
+
+	// We want to allow requests with hosts that don't contain explicit port when the server is running on the  default
+	// port.
+	shouldAllowPortOmitted := false
+	if usingTLS {
+		shouldAllowPortOmitted = allowedPort == "443"
+	} else {
+		shouldAllowPortOmitted = allowedPort == "80"
+	}
+
+	for _, allowedHost := range allowedHosts {
+		allowedMap[fmt.Sprintf("%v:%v", allowedHost, allowedPort)] = true
+		// When appropriate, create entry in allowedMap without explicit port
+		if shouldAllowPortOmitted {
+			allowedMap[allowedHost] = true
+		}
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			host := requestHost(r)
+			if allowedMap[host] {
```

Comment:
This is an exact match which might be too restrictive. Looking at other libraries they tend to support wildcards and/or suffix matching e.g. `".example.com"` or `"*.example.com"`.

Examples:
- https://www.playframework.com/documentation/2.8.x/AllowedHostsFilter
- https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.hostfiltering.hostfilteringoptions.allowedhosts?view=aspnetcore-3.1
- https://docs.djangoproject.com/en/3.0/ref/settings/#allowed-hosts